persist/fs/files.go

```diff
@@ -211,29 +342,46 @@ func forEachInfoFile(filePathPrefix string, namespace ident.ID, shard uint32, re
 		if err != nil {
 			continue
 		}
-		fn(matched[i], infoData)
+		if len(matched[i].AbsoluteFilepaths) != 1 {
+			continue
+		}
+
+		fn(matched[i].AbsoluteFilepaths[0], infoData)
 	}
 }
 
-// ReadInfoFiles reads all the valid info entries.
+// ReadInfoFileResult is the result of reading an info file
+type ReadInfoFileResult struct {
+	Info schema.IndexInfo
+	Err  error
+}
+
+// ReadInfoFiles reads all the valid info entries. Even if ReadInfoFiles returns an error,
+// there may be some valid entries in the returned slice.
 func ReadInfoFiles(
 	filePathPrefix string,
 	namespace ident.ID,
 	shard uint32,
 	readerBufferSize int,
 	decodingOpts msgpack.DecodingOptions,
-) []schema.IndexInfo {
-	var indexEntries []schema.IndexInfo
+) []ReadInfoFileResult {
+	var infoFileResults []ReadInfoFileResult
 	decoder := msgpack.NewDecoder(decodingOpts)
 	forEachInfoFile(filePathPrefix, namespace, shard, readerBufferSize, func(_ string, data []byte) {
 		decoder.Reset(msgpack.NewDecoderStream(data))
 		info, err := decoder.DecodeIndexInfo()
-		if err != nil {
-			return
-		}
-		indexEntries = append(indexEntries, info)
+		infoFileResults = append(infoFileResults, ReadInfoFileResult{
```

Comment:
one more nit about this: `info` won't be set if `err != nil`, so this doesn't give the caller sufficient information to identify which info file is fked.

Reply:
You could add a filepath field to `ReadInfoFileResult` or even:

Reply:
done